File: go/sync_service.go
// sync_service.go
//
// Ports CircleAI.Sync.IMemorySyncService (IMemorySyncService.cs) and
// CircleAI.Sync.MemorySyncService (MemorySyncService.cs) — the push/receive
// orchestrator that serialises memory deltas, routes them through an
// ISyncChannel, and applies received deltas to the local IEpisodicMemoryStore.
//
// Also ports CircleAI.Sync.SyncPrimitives (SyncPrimitives.cs): VersionVector +
// SyncReconciliation (version-vector merge, dominance, last-writer-wins).
//
// The C# reference leaves the episodic apply as a placeholder comment. Per the
// "no stubs" rule the receive loop here is fully wired: episodic-domain deltas
// carry a JSON-encoded EpisodicMemoryEntry (see EncodeEpisodicDelta) that is
// deserialised and upserted into the local store.

package circleai

import (
	"context"
	"encoding/json"
	"errors"
	"sync"
	"time"

	"github.com/google/uuid"
)

// ─────────────────────────────────────────────────────────────────────────────
// IMemorySyncService + MemorySyncService
// ─────────────────────────────────────────────────────────────────────────────

// IMemorySyncService pushes and receives memory deltas across all owned devices.
// The transport is determined by ISyncChannel — app code is identical whether
// the delta travels gRPC, BLE mesh, or DTN bundle.
type IMemorySyncService interface {
	// PushMemoryDelta pushes a memory delta for ownerID to all other devices.
	PushMemoryDelta(ctx context.Context, ownerID, domainKey string, delta []byte, mode SyncDeliveryMode) error

	// StartReceiving starts receiving and applying incoming deltas for ownerID.
	StartReceiving(ctx context.Context, ownerID string) error

	// StopReceiving stops receiving.
	StopReceiving(ctx context.Context) error
}

// MemorySyncService is the default IMemorySyncService implementation.
type MemorySyncService struct {
	channel       ISyncChannel
	store         IEpisodicMemoryStore
	localDeviceID string
	now           func() time.Time

	mu         sync.Mutex
	receiveCtx context.Context
	receiveFn  context.CancelFunc
	wg         sync.WaitGroup
}

// NewMemorySyncService wires a service over a sync channel, episodic store, and
// this device's id. All three are required.
func NewMemorySyncService(channel ISyncChannel, store IEpisodicMemoryStore, localDeviceID string) (*MemorySyncService, error) {
	if channel == nil {
		return nil, errors.New("channel required")
	}
	if store == nil {
		return nil, errors.New("store required")
	}
	if isBlank(localDeviceID) {
		return nil, errors.New("localDeviceId required")
	}
	return &MemorySyncService{
		channel:       channel,
		store:         store,
		localDeviceID: localDeviceID,
		now:           func() time.Time { return time.Now().UTC() },
	}, nil
}

// PushMemoryDelta builds a broadcast SyncDelta and pushes it via the channel.
// The Sequence is the current Unix-ms timestamp, mirroring the C# reference.
func (s *MemorySyncService) PushMemoryDelta(ctx context.Context, ownerID, domainKey string, delta []byte, mode SyncDeliveryMode) error {
	syncDelta := SyncDelta{
		OwnerID:        ownerID,
		SourceDeviceID: s.localDeviceID,
		TargetDeviceID: "", // broadcast to all owned devices
		DomainKey:      domainKey,
		Payload:        delta,
		Sequence:       s.now().UnixMilli(),
		DeliveryMode:   mode,
		TTL:            nil,
		CreatedAt:      s.now(),
	}
	return s.channel.PushDelta(ctx, syncDelta)
}

// StartReceiving launches the receive loop for ownerID. Deltas authored by this
// device are skipped; episodic-domain deltas are applied to the local store.
func (s *MemorySyncService) StartReceiving(ctx context.Context, ownerID string) error {
	if isBlank(ownerID) {
		return errors.New("ownerId required")
	}
	s.mu.Lock()
	if s.receiveFn != nil {
		s.receiveFn()
	}
	rctx, cancel := context.WithCancel(context.Background())
	s.receiveCtx = rctx
	s.receiveFn = cancel
	s.mu.Unlock()

	s.wg.Add(1)
	go func() {
		defer s.wg.Done()
		s.receiveLoop(rctx, ownerID)
	}()
	return nil
}

// StopReceiving cancels the receive loop and waits for it to finish.
func (s *MemorySyncService) StopReceiving(_ context.Context) error {
	s.mu.Lock()
	cancel := s.receiveFn
	s.receiveFn = nil
	s.mu.Unlock()
	if cancel != nil {
		cancel()
	}
	s.wg.Wait()
	return nil
}

func (s *MemorySyncService) receiveLoop(ctx context.Context, ownerID string) {
	deltas, errs := s.channel.ReceiveDeltas(ctx, ownerID, 0)
	for {
		select {
		case <-ctx.Done():
			return
		case <-errs:
			// At most one error then the stream closes; stop the loop.
			return
		case delta, ok := <-deltas:
			if !ok {
				return
			}
			if delta.SourceDeviceID == s.localDeviceID {
				continue // skip own echoes
			}
			if delta.DomainKey == SyncDomainKeys.MemoryEpisodic {
				if entry, decodeErr := DecodeEpisodicDelta(delta.Payload); decodeErr == nil {
					_ = s.store.Add(ctx, entry)
				}
			}
			// Additional domain handlers (affect, persona, goals) go here.
		}
	}
}

var _ IMemorySyncService = (*MemorySyncService)(nil)

// ─────────────────────────────────────────────────────────────────────────────
// Episodic delta wire format
// ─────────────────────────────────────────────────────────────────────────────

// episodicDeltaDTO is the JSON wire shape of an episodic memory delta.
type episodicDeltaDTO struct {
	ID            string            `json:"id"`
	RecordedAtUTC time.Time         `json:"recordedAtUtc"`
	UserText      string            `json:"userText"`
	AssistantText string            `json:"assistantText"`
	AppContext    *string           `json:"appContext,omitempty"`
	Embedding     []float32         `json:"embedding,omitempty"`
	Tags          map[string]string `json:"tags,omitempty"`
}

// EncodeEpisodicDelta serialises an EpisodicMemoryEntry to the episodic-domain
// wire payload used by PushMemoryDelta.
func EncodeEpisodicDelta(entry EpisodicMemoryEntry) ([]byte, error) {
	return json.Marshal(episodicDeltaDTO{
		ID:            entry.ID.String(),
		RecordedAtUTC: entry.RecordedAtUTC,
		UserText:      entry.UserText,
		AssistantText: entry.AssistantText,
		AppContext:    entry.AppContext,
		Embedding:     entry.Embedding,
		Tags:          entry.Tags,
	})
}

// DecodeEpisodicDelta parses an episodic-domain wire payload back into an
// EpisodicMemoryEntry. A blank or invalid id yields a fresh UUID.
func DecodeEpisodicDelta(payload []byte) (EpisodicMemoryEntry, error) {
	var dto episodicDeltaDTO
	if err := json.Unmarshal(payload, &dto); err != nil {
		return EpisodicMemoryEntry{}, err
	}
	id, err := uuid.Parse(dto.ID)
	if err != nil {
		id = uuid.New()
	}
	return EpisodicMemoryEntry{
		ID:            id,
		RecordedAtUTC: dto.RecordedAtUTC,
		UserText:      dto.UserText,
		AssistantText: dto.AssistantText,
		AppContext:    dto.AppContext,
		Embedding:     dto.Embedding,
		Tags:          dto.Tags,
	}, nil
}

// ─────────────────────────────────────────────────────────────────────────────
// SyncPrimitives — VersionVector + SyncReconciliation
// ─────────────────────────────────────────────────────────────────────────────

// VersionVector maps a node id to its logical clock value.
type VersionVector struct {
	// Clocks holds the per-node logical clock values.
	Clocks map[string]int64
}

// NewVersionVector wraps clocks (a nil map is treated as empty).
func NewVersionVector(clocks map[string]int64) VersionVector {
	if clocks == nil {
		clocks = map[string]int64{}
	}
	return VersionVector{Clocks: clocks}
}

// MergeVersionVectors returns the pointwise maximum of a and b — the
// least-upper-bound merge used to reconcile two peers' clocks.
func MergeVersionVectors(a, b VersionVector) VersionVector {
	merged := make(map[string]int64)
	for k, v := range a.Clocks {
		merged[k] = v
	}
	for k, bv := range b.Clocks {
		if av, ok := merged[k]; !ok || bv > av {
			merged[k] = bv
		}
	}
	return VersionVector{Clocks: merged}
}

// VersionVectorADominatesB reports whether a dominates b: a is >= b on every key
// and strictly greater on at least one.
func VersionVectorADominatesB(a, b VersionVector) bool {
	keys := make(map[string]struct{}, len(a.Clocks)+len(b.Clocks))
	for k := range a.Clocks {
		keys[k] = struct{}{}
	}
	for k := range b.Clocks {
		keys[k] = struct{}{}
	}
	anyStrictlyGreater := false
	for k := range keys {
		av := a.Clocks[k]
		bv := b.Clocks[k]
		if av < bv {
			return false
		}
		if av > bv {
			anyStrictlyGreater = true
		}
	}
	return anyStrictlyGreater
}

// LastWriterWins returns the value with the later timestamp; ties favour a
// (a.At >= b.At), matching the C# reference.
func LastWriterWins[T any](aAt time.Time, aVal T, bAt time.Time, bVal T) (time.Time, T) {
	if !aAt.Before(bAt) {
		return aAt, aVal
	}
	return bAt, bVal
}

File: rust/src/sync_service.rs
//! sync_service.rs
//!
//! Port of `CircleAI.Sync/` — the memory-delta push/receive orchestrator.
//!
//!   * [`IMemorySyncService`] — pushes and receives memory deltas across all
//!     owned devices. The transport is [`ISyncChannel`] so the calling code is
//!     identical whether the delta travels gRPC, BLE mesh, or DTN bundle.
//!   * [`MemorySyncService`] — the default implementation. Serialises memory
//!     deltas, routes them through the channel, and applies received deltas to
//!     the local episodic store.
//!
//! The C# `ReceiveLoopAsync` runs as a fire-and-forget background task draining
//! an `IAsyncEnumerable`. The Rust `ISyncChannel::receive_deltas` yields a
//! synchronous iterator, so this port exposes [`MemorySyncService::receive_once`]
//! — drain the currently-available deltas and apply the episodic ones — which a
//! host drives on its own cadence. `start_receiving` / `stop_receiving` gate it
//! via a flag, mirroring the C# `_receiveCts` lifecycle.
//!
//! Applying a delta is a concrete operation here (not a TODO): an
//! `EpisodicMemory` delta's JSON payload is deserialised into an
//! [`EpisodicMemoryEntry`] and added to the local store. Deltas whose
//! `source_device_id` equals the local device are skipped (own echoes), as are
//! deltas of other domains.

use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::Mutex;

use crate::memory::stores::{EpisodicMemoryEntry, IEpisodicMemoryStore};
use crate::sync::{ISyncChannel, SyncDeliveryMode, SyncDelta, SyncDomainKeys};

// ─────────────────────────────────────────────────────────────────────────────
// IMemorySyncService
// ─────────────────────────────────────────────────────────────────────────────

/// Pushes and receives memory deltas across all owned devices.
pub trait IMemorySyncService {
    type Error: std::error::Error;

    /// Push a memory delta for `owner_id` to all other devices.
    fn push_memory_delta(
        &self,
        owner_id: &str,
        domain_key: &str,
        delta: Vec<u8>,
        mode: SyncDeliveryMode,
    ) -> Result<(), Self::Error>;

    /// Start receiving and applying incoming deltas for `owner_id`.
    fn start_receiving(&self, owner_id: &str) -> Result<(), Self::Error>;

    /// Stop receiving.
    fn stop_receiving(&self) -> Result<(), Self::Error>;
}

// ─────────────────────────────────────────────────────────────────────────────
// MemorySyncService
// ─────────────────────────────────────────────────────────────────────────────

/// Default [`IMemorySyncService`]. Serialises memory deltas, routes them through
/// an [`ISyncChannel`], and applies received episodic deltas to the local store.
///
/// Generic over the channel `C` and the episodic store `S` (both sync traits use
/// an associated `Error`, so monomorphising keeps the port allocation-free and
/// object-safe). 1:1 with the C# `MemorySyncService`.
pub struct MemorySyncService<C, S>
where
    C: ISyncChannel,
    S: IEpisodicMemoryStore,
{
    channel: Mutex<C>,
    store: Mutex<S>,
    local_device_id: String,
    receiving: AtomicBool,
    receiving_owner: Mutex<Option<String>>,
}

/// Error surface for [`MemorySyncService`]. Wraps either the channel's or the
/// store's error, plus payload-decode failures.
#[derive(Debug)]
pub enum MemorySyncError<CE, SE> {
    /// The underlying [`ISyncChannel`] failed.
    Channel(CE),
    /// The underlying [`IEpisodicMemoryStore`] failed.
    Store(SE),
}

impl<CE: std::fmt::Display, SE: std::fmt::Display> std::fmt::Display for MemorySyncError<CE, SE> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            MemorySyncError::Channel(e) => write!(f, "sync channel error: {e}"),
            MemorySyncError::Store(e) => write!(f, "episodic store error: {e}"),
        }
    }
}

impl<CE, SE> std::error::Error for MemorySyncError<CE, SE>
where
    CE: std::fmt::Debug + std::fmt::Display,
    SE: std::fmt::Debug + std::fmt::Display,
{
}

impl<C, S> MemorySyncService<C, S>
where
    C: ISyncChannel,
    S: IEpisodicMemoryStore,
{
    /// Creates the service over a channel, a local episodic store, and this
    /// device's id.
    pub fn new(channel: C, store: S, local_device_id: impl Into<String>) -> Self {
        Self {
            channel: Mutex::new(channel),
            store: Mutex::new(store),
            local_device_id: local_device_id.into(),
            receiving: AtomicBool::new(false),
            receiving_owner: Mutex::new(None),
        }
    }

    /// Whether a receive session is currently active.
    pub fn is_receiving(&self) -> bool {
        self.receiving.load(Ordering::SeqCst)
    }

    /// Drains the deltas currently available for `owner_id` and applies the
    /// episodic-memory ones to the local store. Own-device echoes and
    /// other-domain deltas are skipped. No-op (returns `Ok(0)`) unless a receive
    /// session for this owner is active (see [`start_receiving`]).
    ///
    /// Returns the number of episodic deltas actually applied. This is the
    /// synchronous analog of the C# `ReceiveLoopAsync` body, invoked by the host
    /// on its own cadence.
    pub fn receive_once(&self, owner_id: &str) -> Result<usize, MemorySyncError<C::Error, S::Error>> {
        if !self.is_receiving() {
            return Ok(0);
        }
        {
            let active = self.receiving_owner.lock().unwrap();
            if active.as_deref() != Some(owner_id) {
                return Ok(0);
            }
        }

        // Collect first so the channel lock is released before we touch the store
        // (mirrors the C# loop which awaits per-item but never holds two locks).
        let deltas: Vec<SyncDelta> = {
            let channel = self.channel.lock().unwrap();
            let iter = channel
                .receive_deltas(owner_id)
                .map_err(MemorySyncError::Channel)?;
            let mut out = Vec::new();
            for item in iter {
                out.push(item.map_err(MemorySyncError::Channel)?);
            }
            out
        };

        let mut applied = 0usize;
        for delta in deltas {
            if delta.source_device_id == self.local_device_id {
                continue; // skip own echoes
            }
            if delta.domain_key == SyncDomainKeys::MEMORY_EPISODIC {
                // Full wire: deserialise and upsert into the local episodic store.
                if let Ok(entry) = serde_json::from_slice::<EpisodicMemoryEntry>(&delta.payload) {
                    let mut store = self.store.lock().unwrap();
                    store.add(entry).map_err(MemorySyncError::Store)?;
                    applied += 1;
                }
            }
            // Additional domain handlers (affect, persona, goals) go here.
        }
        Ok(applied)
    }
}

impl<C, S> IMemorySyncService for MemorySyncService<C, S>
where
    C: ISyncChannel,
    S: IEpisodicMemoryStore,
{
    type Error = MemorySyncError<C::Error, S::Error>;

    fn push_memory_delta(
        &self,
        owner_id: &str,
        domain_key: &str,
        delta: Vec<u8>,
        mode: SyncDeliveryMode,
    ) -> Result<(), Self::Error> {
        let sync_delta = SyncDelta::new(
            owner_id,
            &self.local_device_id,
            "", // broadcast to all owned devices
            domain_key,
            delta,
            chrono::Utc::now().timestamp_millis(),
            mode,
            None, // ttl
        );

        let mut channel = self.channel.lock().unwrap();
        channel
            .push_delta(&sync_delta)
            .map_err(MemorySyncError::Channel)
    }

    fn start_receiving(&self, owner_id: &str) -> Result<(), Self::Error> {
        *self.receiving_owner.lock().unwrap() = Some(owner_id.to_string());
        self.receiving.store(true, Ordering::SeqCst);
        Ok(())
    }

    fn stop_receiving(&self) -> Result<(), Self::Error> {
        self.receiving.store(false, Ordering::SeqCst);
        *self.receiving_owner.lock().unwrap() = None;
        Ok(())
    }
}
